tests/components/rest/test_switch.py

```diff
@@ -96,286 +121,339 @@ async def test_setup_query_params(
-            rest.CONF_STATE_RESOURCE: STATE_RESOURCE,
-            rest.CONF_METHOD: METHOD,
-            rest.CONF_HEADERS: headers,
-            rest.CONF_PARAMS: PARAMS,
-            rest.CONF_BODY_ON: body_on,
-            rest.CONF_BODY_OFF: body_off,
-            rest.CONF_IS_ON_TEMPLATE: None,
-            rest.CONF_TIMEOUT: 10,
-            rest.CONF_VERIFY_SSL: True,
-        },
-        None,
-    )
-    switch.hass = hass
-    return switch, body_on, body_off
+async def _async_setup_test_switch(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    aioclient_mock.get(RESOURCE, status=HTTPStatus.OK)
 
+    headers = {"Content-type": CONTENT_TYPE_JSON}
+    config = {
+        CONF_PLATFORM: DOMAIN,
+        CONF_NAME: NAME,
+        CONF_DEVICE_CLASS: DEVICE_CLASS,
+        CONF_RESOURCE: RESOURCE,
+        CONF_STATE_RESOURCE: STATE_RESOURCE,
+        CONF_HEADERS: headers,
+    }
 
-def test_name(hass: HomeAssistant) -> None:
-    """Test the name."""
-    switch, body_on, body_off = _setup_test_switch(hass)
-    assert switch.name == NAME
+    assert await async_setup_component(hass, SWITCH_DOMAIN, {SWITCH_DOMAIN: config})
+    await hass.async_block_till_done()
+    assert_setup_component(1, SWITCH_DOMAIN)
 
+    assert hass.states.get("switch.foo").state == STATE_UNKNOWN
+    aioclient_mock.clear_requests()
 
-def test_device_class(hass: HomeAssistant) -> None:
+
+async def test_name(hass: HomeAssistant, aioclient_mock: AiohttpClientMocker) -> None:
     """Test the name."""
-    switch, body_on, body_off = _setup_test_switch(hass)
-    assert switch.device_class == DEVICE_CLASS
+    await _async_setup_test_switch(hass, aioclient_mock)
+
+    state = hass.states.get("switch.foo")
+    assert state.attributes[ATTR_FRIENDLY_NAME] == NAME
+
+
+async def test_device_class(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test the device class."""
+    await _async_setup_test_switch(hass, aioclient_mock)
 
+    state = hass.states.get("switch.foo")
+    assert state.attributes[ATTR_DEVICE_CLASS] == DEVICE_CLASS
 
-def test_is_on_before_update(hass: HomeAssistant) -> None:
+
+async def test_is_on_before_update(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
     """Test is_on in initial state."""
-    switch, body_on, body_off = _setup_test_switch(hass)
-    assert switch.is_on is None
+    await _async_setup_test_switch(hass, aioclient_mock)
+
+    state = hass.states.get("switch.foo")
+    assert state.state == STATE_UNKNOWN
 
 
 async def test_turn_on_success(
     hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
 ) -> None:
     """Test turn_on."""
-    aioclient_mock.post(RESOURCE, status=HTTPStatus.OK)
-    switch, body_on, body_off = _setup_test_switch(hass)
-    await switch.async_turn_on()
+    await _async_setup_test_switch(hass, aioclient_mock)
+
+    with patch("homeassistant.components.rest.switch.RestSwitch.async_update"):
```

Comment:
We shouldn't patch an entity method. What is the problem that we want to solve?

Reply:
What is the problem with triggering `async_update`?

Reply:
If we trigger `async_update`, then we are making two HTTP calls (POST + GET), and the state comes from the GET.

In the old tests, the state was assumed to be set by the call to POST directly.

Since we are not calling `async_update_ha_state`, is it correct that calling `TURN_ON` automatically generates a state update?

Reply:
Note: I am not sure if the issue comes from the tests, or from the implementation.
Maybe we just shouldn't set `self._attr_is_on = True/False` on `turn_on/turn_off`

Reply:
> Since we are not calling async_update_ha_state, is it correct that calling TURN_ON automatically generates a state update?

Yes, the entity is polling so will fetch new data for state at the end of the service call.

Reply:
I have worked out why attr_is_on needs to be set on turn_on/turn_off routine: if get_device_state fails, then it uses optimistic mode.
I have adjusted the tests accordingly, to make sure that the two tests use optimistic behavior